server/src/internalClusterTest/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderIT.java

```diff
@@ -0,0 +1,339 @@
+                = (InternalClusterInfoService) internalCluster().getMasterNodeInstance(ClusterInfoService.class);
+        internalCluster().getMasterNodeInstance(ClusterService.class).addListener(event -> clusterInfoService.refresh());
+
+        final String dataNode0Id = internalCluster().getInstance(NodeEnvironment.class, dataNodeName).nodeId();
+        final Path dataNode0Path = internalCluster().getInstance(Environment.class, dataNodeName).dataFiles()[0];
+
+        createIndex("test", Settings.builder()
+                .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+                .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 6)
+                .put(INDEX_STORE_STATS_REFRESH_INTERVAL_SETTING.getKey(), "0ms")
+                .build());
+        final long minShardSize = createReasonableSizedShards();
+
+        // reduce disk size of node 0 so that no shards fit below the high watermark, forcing all shards onto the other data node
+        // (subtract the translog size since the disk threshold decider ignores this and may therefore move the shard back again)
+        fileSystemProvider.getTestFileStore(dataNode0Path).setTotalSpace(minShardSize + WATERMARK_BYTES - 1L);
+        refreshDiskUsage();
+        assertThat(getShardRoutings(dataNode0Id), empty());
+
+        // increase disk size of node 0 to allow just enough room for one shard, and check that it's rebalanced back
+        fileSystemProvider.getTestFileStore(dataNode0Path).setTotalSpace(minShardSize + WATERMARK_BYTES + 1L);
+        refreshDiskUsage();
+        assertThat(getShardRoutings(dataNode0Id), hasSize(1));
+    }
+
+    private Set<ShardRouting> getShardRoutings(String nodeId) {
+        final Set<ShardRouting> shardRoutings = new HashSet<>();
+        for (IndexShardRoutingTable indexShardRoutingTable : client().admin().cluster().prepareState().clear().setRoutingTable(true)
+                .get().getState().getRoutingTable().index("test")) {
+            for (ShardRouting shard : indexShardRoutingTable.shards()) {
+                assertThat(shard.state(), equalTo(ShardRoutingState.STARTED));
+                if (shard.currentNodeId().equals(nodeId)) {
+                    shardRoutings.add(shard);
+                }
+            }
+        }
+        return shardRoutings;
+    }
+
+    /**
+     * Index documents until all the shards are at least WATERMARK_BYTES in size, and return the size of the smallest shard
+     */
+    private long createReasonableSizedShards() throws InterruptedException {
+        while (true) {
+            final IndexRequestBuilder[] indexRequestBuilders = new IndexRequestBuilder[scaledRandomIntBetween(100, 10000)];
+            for (int i = 0; i < indexRequestBuilders.length; i++) {
+                indexRequestBuilders[i] = client().prepareIndex("test").setSource("field", randomAlphaOfLength(10));
+            }
+            indexRandom(true, indexRequestBuilders);
+            forceMerge();
```

Comment:
why the force-merge?

Reply:
seems not to work otherwise. Not really relevant whether or not we  force-merge, so I left as is